src/MongoDB.Driver.Core/Core/Authentication/AwsSignatureVersion4.cs

```diff
@@ -0,0 +1,203 @@
+﻿/* Copyright 2020–present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace MongoDB.Driver.Core.Authentication
+{
+    /// <summary>
+    /// The AWS signature version 4.
+    /// </summary>
+    public class AwsSignatureVersion4
+    {
+        /// <summary>
+        /// Gets canonical headers.
```

Comment:
`GetCanonicalHeaders` along with `GetRegion` and `GetSignedHeaders` were made public to allow unit testing them, as those are important parts of functionality.

Reply:
Could you make these methods protected instead? I'm not sure _exactly_ how this works in C# but it looks almost identical to Java. We were able to unit test these methods while leaving them protected so as not to expose new public methods to users.

Reply:
I don't think that we need to make these methods public only for test reasons. They could be simply called via reflection